### ISSUE TITLE
Unable to use CASE statement and variable in ORDER BY

### ISSUE THREAD

**Issue by damonLL** [precog/quasar#2255]
SlamData 4.2.1
Quasar 16.1.9
MongoDB 3.4

Customer has a defined MarkDown variable `:SortBy`. It is a radio button which is used in a query in a subsequent Query Card:

```
SELECT
  o1, o2, o3
FROM `/mydb/c1`
ORDER BY
(CASE :SortBy
  WHEN "o1-asc" THEN o1 `ASC`
  WHEN "o2-asc" THEN o2 `ASC`
  WHEN "o3-asc" THEN o3 `ASC`
END)
```
When executed, the following error is received:
```
{
  "error": {
    "status": "Malformed SQL^2 query.",
    "detail": {
      "message": "keyword 'end' expected; but found `ASC'"
    }
  }
}
```
Attempted to try several different `CASE` approaches, including wrapping the ``` o1 `ASC` ``` in quotes, removing back-ticks from keywords, etc.

The JS Console displays the following query being sent to Quasar from SlamData:

```
q:SELECT
  o1, o2, o3
FROM `/mydb/c1`
ORDER BY
(CASE :SortBy
  WHEN "o1-asc" THEN o1 `ASC`
  WHEN "o2-asc" THEN o2 `ASC`
  WHEN "o3-asc" THEN o3 `ASC`
END)
var.SortBy:"o2-asc"
```

This seems like a bug, unless the functionality was not intended. Original use case was to allow users (not authors) in SlamData to choose which column to sort on, so the customer wanted to provide a work around. Neither our Preview Table nor the Pivot Table provides this functionality, so I will open a feature request for that shortly.

**Comment by zeeiyerWork** [comment]
@sellout - this is affecting a customer, so can we look at this with priority(hence in ASAP milestone). If it is not an old-mongo connector issue, then feel free to reassign. also let us know if it will work in new-mongo.

**Comment by sellout** [comment]
The query looks incorrect to me – `o1 asc` is not an expression. `asc` only makes sense as part of the `order by` operation. Try
```sql
SELECT
  o1, o2, o3
FROM `/mydb/c1`
ORDER BY
(CASE :SortBy
  WHEN "o1-asc" THEN o1
  WHEN "o2-asc" THEN o2
  WHEN "o3-asc" THEN o3
END) ASC
```

**Comment by damonLL** [comment]
Actually customer would like the user to be able to control ASC or DESC as well. I suppose a second CASE statement would work? I'll try shortly. Thanks @sellout!

**Comment by sellout** [comment]
@damonLL It won’t unfortunately, because `asc`/`desc` isn’t an expression. We could discuss how to make that work. Currently, the only option (I think, haven’t tried it) is like
```sql
case :SortBy
  when "o1-asc" then (select o1, o2, o3 from `/mydb/c1` order by o1 asc)
  when "o1-desc" then (select o1, o2, o3 from `/mydb/c1` order by o1 desc)
end
```
Where we’d like to get to with a totally expression-based SQL² is
```sql
select o1, o2, o2 (case :SortBy
  when "o1-asc" then (from `/mydb/c1` order by o1 asc)
  when "o1-desc" then (from `/mydb/c1` order by o1 desc)
end)
```
There has also been talk of higher-order functions, which would allow something like
```sql
select o1, o2, o2 apply(case :SortBy
  when "o1-asc" then (_ order by o1 asc)
  when "o1-desc" then (_ order by o1 desc)
end, from `/mydb/c1`)
```
but I’m not even sure if that’s the syntax that would come out of that.

**Comment by sellout** [comment]
Oh, so you could do it with two cases, like:
```sql
case
  when :SortBy like "%-asc"
  then (select o1, o2, o3 from `/mydb/c1` order by (case
      when :SortBy like "o1-%" then o1
      when :SortBy like "o2-%" then o2
      when :SortBy like "o3-%" then o3
    end) asc)
  else (select o1, o2, o3 from `/mydb/c1` order by (case
      when :SortBy like "o1-%" then o1
      when :SortBy like "o2-%" then o2
      when :SortBy like "o3-%" then o3
    end) desc)
end
```
Again, I haven’t tried any of this. And that inner case could probably be pulled into a `let` binding to eliminate the duplication (but that shouldn’t affect the final query – especially because all of this variable handling is statically-evaluable).

**Comment by sellout** [comment]
Would be nicer (for probably the user as well) if the variable was split into two – `:SortBy` and `SortDir`, so then you can avoid the pattern matching and do
```sql
case :SortDir
  when "asc"
  then (select o1, o2, o3 from `/mydb/c1` order by (case :SortBy
      when "o1" then o1
      when "o2" then o2
      when "o3" then o3
    end) asc)
  else (select o1, o2, o3 from `/mydb/c1` order by (case :SortBy
      when "o1" then o1
      when "o2" then o2
      when "o3" then o3
    end) desc)
end
```

**Comment by sellout** [comment]
@damonLL Let me know if the suggested query (or something similar) works. If not, then there is probably some other ASAP bug to fix, but I don’t think this issue is ASAP anymore.

**Comment by jdegoes** [comment]
@sellout Ideally 'ASC' / 'DESC' would be translated into a parameter to an ordinary sort function, where the type is a sum of two singleton types (e.g. "ascending" / "descending").

**Comment by ocasta** [comment]
The double case works in getting the correct query to run. However, the only way to then display the results is a Preview table with no control over the column order as a pivot table always orders by the first column

**Comment by damonLL** [comment]
@ocasta the Preview Table's column order is essentially based on alpha order, so while this is not ideal, you could potentially alias the column names as "A_field1", "B_field2", etc. and they would appear in the desired order horizontally.

Longer term - I think an approach slightly similar to SlamData Issue [1882](https://github.com/slamdata/slamdata/issues/1882) might be helpful. We want to allow a user to control sorting direction and sort field without going into the Pivot Table configuration which would update the workspace core configuration. I'll open an issue for that if one is not already opened and link it here.

**Comment by damonLL** [comment]
@ocasta: @jdegoes has already created SlamData Issue [1871](https://github.com/slamdata/slamdata/issues/1871). May want to keep an eye on that.

**Comment by rintcius** [comment]
@jdegoes Is everything that is expected to be done for this ticket covered by #2480 ? If that's the case then I think this ticket can be closed in favour of #2480.

**Comment by jdegoes** [comment]
@rintcius I think this ticket should be about rewriting `ORDER BY` to something like `SORT([[a, "ascending"], [b, "descending], ...])`, or at least something non-magical so that a variable can determine the sort order for each sort projection.

Even when #2480 is done, it will not be possible to do, `ORDER BY foo :sort_order` because `ORDER BY ... ASC / DESC` is pure syntax and `ASC` and `DESC` are syntactical elements, not expressions.
